Alternate files with ctrl-6

robfalken opened the pull request:
This is my stab at #7709 

I realize the code is flawed. There's no test coverage, I'm using `clone()` and there are probably better ways to hook into the events. Also, I didn't know what context to use for the keybinding. But maybe with some pointers from someone who actually know what they're doing, I can get this shippable.

Release Notes:

- vim: Added ctrl-6 for [alternate-file](https://vimhelp.org/editing.txt.html#CTRL-%5E) to navigate back and forth between two buffers.


https://github.com/zed-industries/zed/assets/261929/2d10494e-5668-4988-b7b4-417c922d6c61

robfalken commented:
@cla-bot check

robfalken committed:
Added support for alternating between two files

This is my stab at #7709

I realize the code is flawed. There's no test coverage, I'm using clone() and there are probably better ways to hook into the events. Also, I didn't know what context to use for the keybinding. But maybe with some pointers from someone who actually know what they're doing, I can get this shippable.

Release Notes:

Added support for vim's alternate-file functionality to navigate back and forth between two buffers.

JosephTLyons commented on crates/workspace/src/workspace.rs:
```diff
@@ -2510,10 +2513,29 @@ impl Workspace {
+    fn track_alternate_files(&mut self, cx: &mut ViewContext<Self>) {
+        if let Some(item) = self.active_item(cx) {
+            if let Some(project_path) = item.project_path(cx) {
+                let (current, alternative) = self.alternate_file_paths.clone();
```
You can avoid cloning `alternative_file_paths` by borrowing it instead:

```rs
 fn track_alternate_files(&mut self, cx: &mut ViewContext<Self>) {
    if let Some(item) = self.active_item(cx) {
        if let Some(project_path) = item.project_path(cx) {
            let (current, alternative) = &self.alternate_file_paths;
            match current {
                Some(current) => {
                    if *current != project_path {
                        self.alternate_file_paths = (Some(project_path), Some(current.clone()));
                    }
                }
                None => {
                    self.alternate_file_paths = (Some(project_path), None);
                }
            }
        }
    }
}
```

robfalken replied:
Oh, yeah that's better! Thanks, I fixed it now.

robfalken committed:
Borrow instead of cloning `alternative_file_paths`

ConradIrwin commented:
Thanks for this! Looks like the linter is complaining about some unused variables (you can fix that by naming them `_` or similar).

If we want to support this more like vim, we should store the alternate file per-pane instead of per workspace. I think you can re-use much of the approach from here, but move it into the Pane instead (use `pane::AlternateFile`, and register the action there).

If you'd like help getting this over the line, feel free to book time here: https://calendly.com/conradirwin/pairing.

robfalken committed:
Move from workspace to pane context

ConradIrwin committed:
Use existing item if it exists

ConradIrwin commented:
It looks like we just need to check fi the item is already active in this pane, and if so reactivate it. Pushed up a potential fix.

robfalken commented:
Oh, perfect. That's is exactly what I wanted to do, but couldn't figure out how. Thanks!

I updated my screencast in the PR description, and rebased my branch. I can't run the tests, though. I just get this error:

```
error: failed to run custom build command for `live_kit_server v0.1.0 (/Users/rf/Projects/zed/crates/live_kit_server)`

Caused by:
  process didn't exit successfully: `~/Projects/zed/target/debug/build/live_kit_server-2648d7b32afc52c2/build-script-build` (exit status: 101)
  --- stderr
  thread 'main' panicked at crates/live_kit_server/build.rs:5:10:
  called `Result::unwrap()` on an `Err` value: Custom { kind: Other, error: "protoc failed: Could not make proto path relative: protocol/livekit_room.proto: No such file or directory\n" }
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
```

@ConradIrwin I have booked a pairing session at Friday May 10, 2024 ⋅ 11am – 12pm (Mountain Time - Denver) to work on this PR. Is there anything else you'd like to improve, or do you think it's good enough that we should cancel that session?

robfalken committed:
Fix failing tests
